EmailAuth/FirebaseEmailAuthUI/FUIEmailAuth.m

```diff
@@ -190,10 +203,243 @@ - (void)signInWithDefaultValue:(nullable NSString *)defaultValue
 }
 
 - (void)signOut {
+
 }
 
 - (BOOL)handleOpenURL:(NSURL *)URL sourceApplication:(nullable NSString *)sourceApplication {
-  return NO;
+  self.emailLink = URL.absoluteString;
+  
+  // Retrieve continueUrl from URL
+  NSURLComponents *urlComponents = [NSURLComponents componentsWithString:URL.absoluteString];
+  NSString *continueURLString;
+  for (NSURLQueryItem *queryItem in urlComponents.queryItems) {
+    if ([queryItem.name isEqualToString:@"continueUrl"]) {
+      continueURLString = queryItem.value;
+    }
+  }
+  if (!continueURLString) {
+    [FUIAuthBaseViewController showAlertWithMessage:@"Invalid link! Missing continue URL."];
+  }
+
+  // Retrieve url parameters from continueUrl
+  NSMutableDictionary *urlParameterDict= [NSMutableDictionary dictionary];
+  NSURLComponents *continueURLComponents = [NSURLComponents componentsWithString:continueURLString];
+  for (NSURLQueryItem *queryItem in continueURLComponents.queryItems) {
+    urlParameterDict[queryItem.name] = queryItem.value;
+  }
+  // Retrieve parameters from local storage
+  NSMutableDictionary *localParameterDict = [NSMutableDictionary dictionary];
+  localParameterDict[kEmailLinkSignInEmailKey] = [GULUserDefaults.standardUserDefaults stringForKey:kEmailLinkSignInEmailKey];
+  localParameterDict[@"ui_sid"] = [GULUserDefaults.standardUserDefaults stringForKey:@"ui_sid"];
+
+  // Handling flows
+  NSString *urlSessionID = urlParameterDict[@"ui_sid"];
+  NSString *localSessionID = localParameterDict[@"ui_sid"];
+  BOOL sameDevice = urlSessionID && localSessionID && [urlSessionID isEqualToString:localSessionID];
+
+  if (sameDevice) { // Same device
+    if (urlParameterDict[@"ui_pid"] != nil) { // Unverified provider linking
```

Comment:
conditional comment usually goes into the next line. also full stop in the end.  same below

if (foo) {
  // process foo.
  processFoo();
}